test/server/content-charset-test.js

```diff
@@ -0,0 +1,287 @@
+            charsetTestFunc(
+                '/page-with-content-type-header',
+                charsetContentType,
+                getExpectedSrcStr(pageWithMetaSrc, charsetContentType, false),
+                done
+            );
+        });
+
+        it('Should set content charset from meta', function (done) {
+            charsetTestFunc(
+                '/page-with-meta-tag',
+                charsetMeta,
+                getExpectedSrcStr(pageWithMetaSrc, charsetMeta, false),
+                done
+            );
+        });
+
+        it('Should set default content charset', function (done) {
+            charsetTestFunc(
+                '/page-default',
+                charsetDefault,
+                getExpectedSrcStr(pageWithoutMetaSrc, charsetDefault, false),
+                done
+            );
+        });
+    });
+
+    describe('Scripts', function () {
+        var processedScript = scriptProcessor.processResource(scriptSrc);
+        var scriptTestFunc  = function (originUrl, expectedCharset, expectedBody, callback) {
+            var url = urlUtil.getProxyUrl('http://127.0.0.1:2000' +
+                                          originUrl, proxyHostname, proxyPort1, sessionId, urlUtil.SCRIPT, expectedCharset);
+
+            request(url, function (err, res, body) {
+                compareCode(body, expectedBody);
+
+                callback();
+            });
+        };
+
+        it('Should set content charset from BOM', function (done) {
+            scriptTestFunc(
+                '/script-with-bom',
+                charsetBOM,
+                iconv.encode(processedScript, charsetBOM, { addBOM: true }).toString(),
+                done
+            );
+        });
+
+        it('Should set content charset from Content-Type header', function (done) {
+            scriptTestFunc(
+                '/script-with-content-type-header',
+                charsetContentType,
+                iconv.encode(processedScript, charsetContentType).toString(),
+                done
+            );
+        });
+
+        it('Should set content charset from url', function (done) {
+            scriptTestFunc(
+                '/script-with-charset-in-url',
+                charsetUrl,
+                iconv.encode(processedScript, charsetUrl).toString(),
+                done
+            );
+        });
+    });
+
+    describe('Other resources', function () {
+        var urlReplacer           = function (url) {
+            return urlUtil.getProxyUrl(url, proxyHostname, proxyPort1, sessionId);
+        };
+        var testOtherResourceFunc = function (resourceType, expectedBody, callback) {
+            var resourceUrl = 'http://127.0.0.1:2000/other-resource/' + resourceType + '/';
+            var proxyUrl    = urlUtil.getProxyUrl(resourceUrl + 'bom', proxyHostname, proxyPort1, sessionId);
+
+            request(proxyUrl, function (err, res, body) {
```

Comment:
Promisify request and use promises instead of nested callbacks (seems like you can test url in parallel).